wolfcrypt/src/chacha.c

```diff
@@ -265,15 +267,40 @@ static void wc_Chacha_encrypt_bytes(ChaCha* ctx, const byte* m, byte* c,
 
     output = (byte*)temp;
 
+    /* handle left overs */
+    if (ctx->left > 0) {
+        wc_Chacha_wordtobyte(temp, ctx->X); /* recreate the stream */
+        output = (byte*)temp + CHACHA_CHUNK_BYTES - ctx->left;
+        for (i = 0; i < bytes && i < ctx->left; i++) {
+            c[i] = m[i] ^ output[i];
+        }
+        ctx->left = ctx->left - i;
+
+        /* Used up all of the stream that was left, increment the counter */
+        if (ctx->left == 0) {
+            ctx->X[CHACHA_IV_BYTES] = PLUSONE(ctx->X[CHACHA_IV_BYTES]);
+        }
+        bytes = bytes - i;
+        c += i;
+        m += i;
+        output = (byte*)temp;
+    }
+
     for (; bytes > 0;) {
         wc_Chacha_wordtobyte(temp, ctx->X);
-        ctx->X[CHACHA_IV_BYTES] = PLUSONE(ctx->X[CHACHA_IV_BYTES]);
         if (bytes <= CHACHA_CHUNK_BYTES) {
             for (i = 0; i < bytes; ++i) {
                 c[i] = m[i] ^ output[i];
             }
+            ctx->left = CHACHA_CHUNK_BYTES - i;
+
+            /* increment the counter if no more bytes are left */
+            if (ctx->left == 0) {
+                ctx->X[CHACHA_IV_BYTES] = PLUSONE(ctx->X[CHACHA_IV_BYTES]);
```

Comment:
Are you sure this is the right spot? What about this case but then 64 bytes are wanted on the next call. If you look below the plus one is called again before any bytes are consumed.

Reply:
TLDR: pretty sure, can add some test cases though if wanted.

The counter should be incremented "at least" before the next 64 byte stream needs to be created. The behavior before this PR was to increment the counter in preparation for the next call to wc_Chacha_wordtobyte immediately after creating a 64 byte stream block. Now in the case that the bytes left is equal to 64 or less it needs to keep ctx->X the same value, no longer incrementing it immediately in preparation for the next call. There is the case though that the last block is equal to 64 and in that case ctx->left will get set to 0 so before hitting return; the counter needs incremented by 1 in preparation for the next call. On the next call since ctx->left is still at 0 the leftover block will not be hit.

Reply:
Please consider the case where the users asks for 64 bytes and then 128 bytes.  Line 299 will do plus one at the end of the first call.  But at the start of the 2nd call another plus one will be done on line 303 before any bytes have been consumed since the previous plus one if I'm reading this right.

Reply:
updated the test case in tests/api.c to hit that use case